feat(tianmu): impl bit type, add support to create/desc table with bit type #919

hustjieke opened the pull request:
<!--

Thank you for contributing to StoneDB!

PR Title Format: <type>(<scope>): description to this pr (#issue_id)
e.g.
fix(util): fix sth..... (#1)

-->

## Summary about this PR
<!--

There MUST be one line starting with "Issue Number:  " and 
linking the relevant issues via the "close" or "ref".
e.g.:
Issue: close #1

-->
[summary]
1. add support to create table with bit type
2. add support to desc/show fields with bit type table
3. currently we design bit(M) with (1 <= M <= 63), in the future we'll expand prec to 64.


## Tests Check List
<!-- At least one of next options must be included. -->

- [ ] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog
<!-- At least one of next options must be included. -->

- [x] New Feature
- [ ] Bug Fix
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation
<!-- At least one of next options must be included. -->

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

hustjieke committed:
feat(tianmu): impl bit type, add support to create/desc table with bit type #919

[summary]
1. add support to create table with bit type
2. add support to desc/show fields with bit type table
3. currently we design bit(M) with (1 <= M <= 63), in the future we'll expand prec to 64.

hustjieke committed:
style(tianmu): renaming some variables based on google c++ style #11

hustjieke committed:
feat(tianmu): impl bit type on tianmu engine #919

[summary]
1. impl insert with mode delayed=0/1
2. impl simple select bit data from table

hustjieke committed:
feat(tianmu): support agg on bit types #919

hustjieke committed:
fix ComputeNaturalSize and StaticExtrnalSize on bit calculate, used by desc tables #919

hustjieke committed:
style(tianmu): formatting code #919

lujiashun commented on mysql-test/suite/tianmu/t/bit_type.test:
```diff
@@ -0,0 +1,95 @@
+--source include/have_tianmu.inc
+
```
Please use your own database instead of test database to avoid conflict when parallely running test cases；

hustjieke replied:
Done.

hustjieke committed:
feat(mtr): add bit test db during paralle executing #919

lujiashun reviewed (approved):
LGTM

duanfuxiang0 reviewed (approved):
LGTM

hustjieke committed:
feat(tianmu): impl load bit type #919

[summary]
impl load bit, including empty space and null value, to keep the src code integrity, we add
an bit parse function just like others numeric does.

RingsC commented on storage/tianmu/loader/parsing_strategy.cpp:
```diff
@@ -486,7 +486,7 @@ ParsingStrategy::ParseResult ParsingStrategy::GetOneRow(const char *const buf, s
     }
     String *res = field->str_result(str);
-    DEBUG_ASSERT(res);
+    // DEBUG_ASSERT(res);
```
Why do we comment off this?

hustjieke replied:
This will cause crash when load null data in debug mode, it's fixed by @lujiashun in pr #1153

RingsC commented on storage/tianmu/common/common_definitions.h:
```diff
@@ -138,6 +139,8 @@ constexpr uint64_t TIANMU_BIGINT_UNSIGNED_MAX = 0xFFFFFFFFFFFFFFFFULL;  // 2^64
 constexpr int32_t TIANMU_MAX_INDEX_COL_LEN_LARGE = 3072;
 constexpr int32_t TIANMU_MAX_INDEX_COL_LEN_SMALL = 767;
 
+constexpr uint32_t kTianmuBitMaxPrec = 63;  // in the future we'll expand to 64.
```
Pls, use the same naming convention as the above does, Thanks.

hustjieke replied:
For the constant var names, I think we should follow google c++ style:
https://google.github.io/styleguide/cppguide.html#Constant_Names
```
const int kDaysInAWeek = 7;
const int kAndroid8_0_0 = 24;  // Android 8.0.0
```
There are a lot of const vars defined in `tianmu` is not specification, and I want to uniform all of them in the future.

hustjieke replied:
Also, the `enum` var has the same problem:
https://google.github.io/styleguide/cppguide.html#Enumerator_Names
```
enum class UrlTableError {
  kOk = 0,
  kOutOfMemory,
  kMalformedInput,
};
```

RingsC commented on mysql-test/suite/tianmu/t/create_table.test:
```diff
@@ -397,6 +397,24 @@ show create table t1;
+ bit8 bit(8),
+ bit16 bit(16),
+ bit32 bit(32),
+ bit63 bit(63)
```
adds cases for testing bit(64).  corner test cases

RingsC replied:
Out-of-Range and Overflow Handling
ref: https://dev.mysql.com/doc/refman/8.0/en/out-of-range-and-overflow.html

hustjieke replied:
> adds cases for testing bit(64). corner test cases

`bit(64)` has not been supported yet, I write fail test case when create bit with 64:
```
--error 6
create table tc(a bit(64)) engine=tianmu;
```

RingsC commented on mysql-test/suite/tianmu/t/bit_type.test:
```diff
@@ -0,0 +1,123 @@
+INSERT INTO t SET b = B'1010';
+INSERT INTO t SET b = 0b0101;
+INSERT INTO t values(b'1'), (B'1010'), (0b0101);
+SELECT b+0, BIN(b), OCT(b), HEX(b) FROM t;
```
adds cases for `Illegal bit-value literals`

hustjieke replied:
> adds cases for `Illegal bit-value literals`

I'll add these cases during insert values.

RingsC commented on mysql-test/suite/tianmu/t/bit_type.test:
```diff
@@ -0,0 +1,123 @@
+# test string ... : https://dev.mysql.com/doc/refman/5.7/en/string-comparison-functions.html
+select a like "10111", b not like "1011" from bit2;
+select strcmp(a,b), strcmp(b,a),strcmp(a,a) from bit2;
+select charset(binary a), collation(binary b) from bit2;
```
adds case for utf8 collation
```
SELECT _latin1 b'1000001';
SELECT _utf8mb4 0b1000001 COLLATE utf8mb4_danish_ci;
```
A leading 0b is case-sensitive and cannot be written as `0B`.

hustjieke replied:
Yes, added.

RingsC commented on storage/tianmu/common/txt_data_format.cpp:
```diff
@@ -51,9 +53,9 @@ uint TxtDataFormat::StaticExtrnalSize(ColumnType attrt, uint precision, int scal
-    return 11;
-  else if (attrt == ColumnType::BIGINT)
-    return 20;
+    return 11;  // -2,147,483,648 ~ 2,147,483,647, 1(signed) + 10(max digits)
```
Sugg: DO NOT USE MAGIC NUMBER DIRECTLY, while using meaningful literals. Refactor this part.Thanks.

RingsC replied:
```
A bit-value literal may have an optional character set introducer and COLLATE clause, to designate it as a string that uses a particular character set and collation:

[_charset_name] b'val' [COLLATE collation_name]
```
Therefore, the charset and collation should be considered, if we specifed the charset and collation.

RingsC replied:
Character Set Introducers:
https://dev.mysql.com/doc/refman/8.0/en/charset-introducer.html

RingsC commented on storage/tianmu/core/column.h:
```diff
@@ -24,7 +24,7 @@
 namespace Tianmu {
 namespace core {
-enum class PackOntologicalStatus { NULLS_ONLY, UNIFORM, UNIFORM_AND_NULLS, SEQUENTIAL, NORMAL };
+enum class PackOntologicalStatus { kNullsOnly = 0, kUniform, kUniformAndNulls, kSequential, kNormal };
```
keep the old one, using the same naming convention in our project.

RingsC commented on storage/tianmu/core/column_type.h:
```diff
@@ -27,9 +27,9 @@ struct DataType;
-    BLOOM_FILTER = 2,
+    kNotNull = 0,
+    kAutoInc,
+    kBloomFilter,
```
Keep the old one.

RingsC commented on storage/tianmu/core/data_type.cpp:
```diff
@@ -53,6 +53,11 @@ DataType::DataType(common::ColumnType atype, int prec, int scale, DTCollation co
+    case common::ColumnType::BIT:
+      DEBUG_ASSERT((prec > 0) && (prec <= common::kTianmuBitMaxPrec));
+      valtype = ValueType::VT_FIXED;
+      fixmax = MAX(TIANMU_TINYINT_MAX, -TIANMU_TINYINT_MIN);  // TODO(fix max value with common::kTianmuBitMaxPrec)
```
why the fixmax is (127,-127) , the definition of bit value is not that?

hustjieke replied:
Fixed.

RingsC commented on storage/tianmu/core/engine.cpp:
```diff
@@ -166,7 +166,7 @@ static std::string has_mem_name(const LEX_STRING &comment) {
 }
 
-bool parameter_equals(THD *thd, enum tianmu_var_name vn, longlong value) {
+bool parameter_equals(THD *thd, enum TianmuVarName vn, longlong value) {
```
Keep the naming style as the others. the `thd` and `value` is low-case, but `TianmuVarName` is uppercase, it seems weird.

hustjieke replied:
Fixed.

RingsC commented on storage/tianmu/core/engine.cpp:
```diff
@@ -487,6 +487,15 @@ void Engine::EncodeRecord(const std::string &table_path, int table_id, Field **f
         *(int64_t *)ptr = v;
         ptr += sizeof(int64_t);
       } break;
+      case MYSQL_TYPE_BIT: {
```
In data_type.cpp, fixmax is defined as `fixmax = MAX(TIANMU_TINYINT_MAX, -TIANMU_TINYINT_MIN);`

hustjieke replied:
Fixed.

RingsC commented on storage/tianmu/core/engine.cpp:
```diff
@@ -487,6 +487,15 @@ void Engine::EncodeRecord(const std::string &table_path, int table_id, Field **f
+        // open it when support M = 64, now all value parsed is < 0.
+        if (v > common::TIANMU_BIGINT_MAX)  // v > bigint max when uint64_t is supported
+          v = common::TIANMU_BIGINT_MAX;    // TODO(fix with bit prec)
+        *(int64_t *)ptr = v;
```
using c++ style type-casting, not c style, Thanks.

hustjieke replied:
Fixed.

RingsC commented on storage/tianmu/core/engine.cpp:
```diff
@@ -636,7 +644,7 @@ AttributeTypeInfo Engine::GetAttrTypeInfo(const Field &field) {
   bool notnull = (field.null_bit == 0);
   auto str = boost::to_upper_copy(std::string(field.comment.str, field.comment.length));
-  bool filter = (str.find("FILTER") != std::string::npos);
+  bool bloom_filter = (str.find("FILTER") != std::string::npos);
```
Here, the keyword `FILTER` means `bloom filter`, or something else.

hustjieke replied:
I will revert it back.

RingsC commented on storage/tianmu/core/engine.h:
```diff
@@ -307,22 +307,22 @@ class ResultExportSender final : public ResultSender {
-  TIANMU_ABORT_ON_THRESHOLD,
-  TIANMU_VAR_LIMIT  // KEEP THIS LAST
+enum class TianmuVarName {
+  kTianmuDataFormat = 0,
```
`kTianmuDataFormat` uses the same naming style.

RingsC commented on storage/tianmu/core/engine.h:
```diff
@@ -307,22 +307,22 @@ class ResultExportSender final : public ResultSender {
-  TIANMU_ABORT_ON_COUNT,
-  TIANMU_ABORT_ON_THRESHOLD,
-  TIANMU_VAR_LIMIT  // KEEP THIS LAST
+enum class TianmuVarName {
```
Sugg: Uses `TianmuParamType` instead of `TianmVarName`.

RingsC commented on storage/tianmu/core/engine_convert.cpp:
```diff
@@ -45,20 +45,20 @@ bool Engine::ConvertToField(Field *field, types::TianmuDataType &rcitem, std::ve
     case MYSQL_TYPE_STRING:
-      if (dynamic_cast<types::BString *>(&rcitem)) {
-        ((types::BString &)rcitem).PutString((char *&)field->ptr, (ushort)field->field_length, false);
+      if (dynamic_cast<types::BString *>(&tianmu_item)) {
```
Using c++ sytle type-casting not c style.

RingsC commented on storage/tianmu/core/pack_str.cpp:
```diff
@@ -755,7 +755,7 @@ void PackStr::LoadUncompressed(system::Stream *f) {
   ASSERT(data_.sum_len == sz, "bad pack! " + std::to_string(data_.sum_len) + "/" + std::to_string(sz));
 }
 
-bool PackStr::Lookup(const types::BString &pattern, uint16_t &id) {
```
Here, Lookup is an lookup operation, not a type or flag. Therefore, `Lookup` is a proper name to identify what we will do in this function.

hustjieke replied:
Yeah, I revert it back to `Lookup`.

RingsC commented on storage/tianmu/core/tianmu_attr.cpp:
```diff
@@ -610,27 +613,27 @@ types::BString TianmuAttr::DecodeValue_S(int64_t code) {
-    types::BString local_rcb = tianmu_n.ToBString();
-    local_rcb.MakePersistent();
-    return local_rcb;
+    types::BString local_tianmu_b = tianmu_n.ToBString();
```
why we do not decleare this var `types::BString local_tianmu_b` at the begiing of this part. Every casese re-define the var. and why do we use `xxx_b`, what the meaning of `_b` is.

hustjieke replied:
> 

Yeah, you are right, change the var name to `tianmu_bstring`, short for `BString`.
But we can't move `    types::BString local_rcb = tianmu_n.ToBString();` to beginning of the function. For each define of `tianmu_bstring`, we should first construct `TianmuNum` or `TianmuDateTime`, then convert them to `tianmu_bstring`.
```
  if (ATI::IsIntegerType(a_type)) {
    types::TianmuNum tianmu_n(code, -1, false, a_type);  // condition false
    types::BString local_tianmu_b = tianmu_n.ToBString();
    ...
  } else if (ATI::IsRealType(a_type)) {
    types::TianmuNum tianmu_n(code, -1, true, a_type);  // condition true
    types::BString local_tianmu_b = tianmu_n.ToBString();
```

RingsC commented on storage/tianmu/core/tianmu_table.cpp:
```diff
@@ -1013,15 +1013,16 @@ int TianmuTable::binlog_insert2load_block(std::vector<loader::ValueCache> &vcs,
-            types::TianmuNum rcd(v, m_attrs[att]->Type().GetScale(), m_attrs[att]->Type().IsFloat(),
-                                 m_attrs[att]->TypeName());
-            s = rcd.ToBString();
+            types::TianmuNum tianmu_d(v, m_attrs[att]->Type().GetScale(), m_attrs[att]->Type().IsFloat(),
```
and here, why var is named with `_d` ?

hustjieke replied:
This is replaced rcd, I made a change from `rcd` to `tianmu_d`.
But the variable `tianmu_d` is ugly, I'll change it to `tianmu_num`.

RingsC commented on storage/tianmu/types/tianmu_num.cpp:
```diff
@@ -58,8 +58,9 @@ TianmuNum &TianmuNum::Assign(int64_t value, short scale, bool is_double, common:
-    if (scale != 0 || attrt == common::ColumnType::UNK) {
+  if (scale != -1 &&
+      !is_double_) {  // check if construct decimal, the UNK is used on temp_table.cpp: GetValueString(..)
+    if (scale != 0 || attrt == common::ColumnType::UNK || attrt == common::ColumnType::BIT) {
```
Why we do add condition of `attrt == common::ColumnType::BIT` here?

hustjieke replied:
> Why we do add condition of `attrt == common::ColumnType::BIT` here?

Emmm, I made a mistake here, the condition should be:
```
if ((scale != 0 && attrt != common::ColumnType::BIT) || attrt == common::ColumnType::UNK) {
```
The condition is used to judge if this type is decimal or not, so when `scale != 0`, it may be `bit` or `decimal`.

hustjieke committed:
feat(tianmu): revert and adjust codes on bit #919

[summary]
1. revert some vars renaming code to keep old style.
2. fix type cast using reinterpret_cast to avoid c-style casting.
3. add more test cases and fix some errors.
4. format code.

RingsC commented on storage/tianmu/core/aggregation_algorithm.cpp:
```diff
@@ -104,7 +104,7 @@ void AggregationAlgorithm::Aggregate(bool just_distinct, int64_t &limit, int64_t
       if (cur_a.term.vc) {
         if (!has_lookup)
-          has_lookup = cur_a.term.vc->Type().IsLookup();
+          has_lookup = cur_a.term.vc->Type().Lookup();
```
the function name in `pack_str.cpp` , `Lookup`, is ok, but here `IsLookup` is proper name. It means different usage. That function is used to look up somethings, and here is a type function, and uses Lookup as its type function is not the `BEST` choice here. 
But recommanded that to use the same name convention:
```
  bool IsLookup() const { return fmt == common::PackFmt::LOOKUP; }
  ColumnType RemovedLookup() const;

  bool IsNumeric() const {
    switch (type) {
      case common::ColumnType::BIN:
      case common::ColumnType::BYTE:
      case common::ColumnType::VARBYTE:
      case common::ColumnType::STRING:
      case common::ColumnType::VARCHAR:
      case common::ColumnType::LONGTEXT:
        return false;
      default:
        return true;
    }
  }

  bool IsKnown() const { return type != common::ColumnType::UNK; };
  bool IsFixed() const { return ATI::IsFixedNumericType(type); };
  bool IsFloat() const { return ATI::IsRealType(type); };
  bool IsInt() const { return IsFixed() && scale == 0; }
  bool IsString() const { return ATI::IsStringType(type); }
  bool IsDateTime() const { return ATI::IsDateTimeType(type); }
  const DTCollation &GetCollation() const { return collation; }
```

hustjieke replied:
Get it! I will fix it in next pull request.

RingsC commented on mysql-test/suite/tianmu/t/bit_type.test:
```diff
@@ -0,0 +1,158 @@
+--error 1406
+insert into bit_test values(b'1', b'1',b'1',b'11111111111111111',b'11',b'111');
+#  Data too long for column 'bit32' at row 1
+--error 1406
```
add some cases to test bit-value type as a condition filter. 
FIY:https://developer.aliyun.com/article/27861

hustjieke replied:
Added.

RingsC commented on storage/tianmu/common/common_definitions.h:
```diff
@@ -138,6 +139,8 @@ constexpr uint64_t TIANMU_BIGINT_UNSIGNED_MAX = 0xFFFFFFFFFFFFFFFFULL;  // 2^64
 constexpr int32_t TIANMU_MAX_INDEX_COL_LEN_LARGE = 3072;
 constexpr int32_t TIANMU_MAX_INDEX_COL_LEN_SMALL = 767;
 
+constexpr uint32_t TIANMU_BIT_MAX_PREC = 63;  // in the future we'll expand to 64.
```
so how do we store the bit-value. just is same as mysqld does in `bit string` or something else. If we store the bit-value in int64? and this var is used for what? if it's used for numeric precision, the precision of `int64` is less than 63. and or in double or numeric. it's a good idea to store bit-value type in these real data type.

hustjieke replied:
It is used to intercept when the bit precision is 64 when createing a table.

RingsC commented on storage/tianmu/common/txt_data_format.cpp:
```diff
@@ -39,7 +39,9 @@ uint TxtDataFormat::StaticExtrnalSize(ColumnType attrt, uint precision, int scal
   // because "-0.1" may be declared as DEC(1,1), so 18 decimal places may take
-  // 21 characters
+  // 21 characters, "-0.xxxxx...", "-", "0", "." need extra 3 bytes, 21 = 3 + 18.
+  if (attrt == ColumnType::BIT)  // unsigned value.
```
This is used by `ColumnType`, and we used int64 to store the bit-value, and as the described by info.

```
//! Number of bytes taken by a value when written out,
//! or number of chars if collation not binary and ColumnType::STRING or ColumnType::VARCHAR
uint TxtDataFormat::StaticExtrnalSize(ColumnType attrt, uint precision, int scale, const DTCollation *col) {
```

should be length of column definition.

hustjieke replied:
You are write, the `StaticExtrnalSize` should be column definition ( for bit, the column lenght is `precision`) these two comments is used by  `ColumnType::NUM` and make some confused, I had moved them to the right place above `if (attrt == ColumnType::NUM)`.

RingsC commented on storage/tianmu/core/data_type.cpp:
```diff
@@ -53,6 +53,12 @@ DataType::DataType(common::ColumnType atype, int prec, int scale, DTCollation co
       fixmax = MAX(TIANMU_TINYINT_MAX, -TIANMU_TINYINT_MIN);
       break;
+    case common::ColumnType::BIT:
+      DEBUG_ASSERT((prec > 0) && (prec <= common::TIANMU_BIT_MAX_PREC));
```
Why do we use this `DEBUG_ASSERT` here, and from the code, can draw that we deal with the `bit-value` type as the `bigint` value type.

hustjieke replied:
Yeah, this line has been deleted to keep consitent with `bigint`.

RingsC commented on storage/tianmu/core/engine.cpp:
```diff
@@ -484,14 +484,23 @@ void Engine::EncodeRecord(const std::string &table_path, int table_id, Field **f
+      } break;
+      case MYSQL_TYPE_BIT: {
+        int64_t v = f->val_int();
+        // DEBUG_ASSERT(v < 0, "bit type data should never less than 0.");
```
why we DO NOT use the logic of `BIGINT`, rather than adding these lines.

hustjieke replied:
Emmm... you are right, I had deleted this useless comment.

RingsC commented on storage/tianmu/core/engine.cpp:
```diff
@@ -710,6 +718,13 @@ AttributeTypeInfo Engine::GetAttrTypeInfo(const Field &field) {
     }
+    case MYSQL_TYPE_BIT: {
+      const Field_bit_as_char *f_bit = ((const Field_bit_as_char *)&field);
+      if (/*f_bit->field_length > 0 && */ f_bit->field_length <= common::TIANMU_BIT_MAX_PREC)
```
here, we store the bit-value type in int64, and here why we throw this unspported exception.  as mysql does, it deals with bit-value type as bit string, so that the M should be in [1,64].

hustjieke replied:
As bit type is an unsigned numeric, in fact we should use `uint64_t` to store bit. but now the unsigned `int64_t` has not been supported, we use `int64_t` instead, thus the precision limited to [1, 63], we'll support `M=64`  in next milestone.

hustjieke committed:
feat(tianmu): rm some useless comment, bit type judge and add select...where test case on bit types #919
